### PR TITLE
Nighthawk skeleton

### PR DESCRIPTION
**Pull request by oschaaf**
To make reviewing easier we will be splitting https://github.com/envoyproxy/envoy-perf/pull/38 into smaller, logical chunks. This is the first one, which has most of the bootstrap plus a simple 'Frequency' class so we can check CI is wired up OK.

### CODE REVIEW

**Comment by oschaaf** [comment]
@htuch wdyt, would this be a good first step to try and land?

File: nighthawk/BUILD [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+load(
+    "@envoy//bazel:envoy_build_system.bzl",
+    "envoy_cc_binary",
+)
+
+envoy_cc_binary(
+    name = "nighthawk_client",
+    repository = "@envoy",
+    stamped = True,
```
Comment (htuch) [inline nighthawk/BUILD]: How come you turned this on?

File: nighthawk/WORKSPACE [hunk 2]
```diff
@@ -0,0 +1,30 @@
+workspace(name = "nighthawk")
```
Comment (htuch) [inline nighthawk/WORKSPACE]: Ultimately, I think we might want a root WORKSPACE, Bazel prefers to see each repo as a single tree.

File: nighthawk/ci/do_ci.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash -e
```
Comment (htuch) [inline nighthawk/ci/do_ci.sh]: How come we have this as well as the root `ci/do_ci.sh?`

File: nighthawk/include/nighthawk/client/BUILD [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+    "envoy_package",
+)
+
+licenses(["notice"])  # Apache 2
+
+envoy_package()
+
+envoy_basic_cc_library(
+    name = "client_includes",
+    hdrs = [
```
Comment (htuch) [inline nighthawk/include/nighthawk/client/BUILD]: ?

File: nighthawk/include/nighthawk/common/exception.h [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+ * help is invoked. This exception is then caught further up the stack and handled.
+ *
+ */
+class NoServingException : public NighthawkException {
+public:
+  NoServingException() : NighthawkException("NoServingException") {}
+};
+
+/**
+ * @brief Thrown when an OptionsImpl was not constructed because the argv was invalid.
```
Comment (htuch) [inline nighthawk/include/nighthawk/common/exception.h]: We don't use `@brief` typically in Envoy.

File: nighthawk/include/nighthawk/common/exception.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  NighthawkException(const std::string& message) : std::runtime_error(message) {}
+};
+
+// TODO(oschaaf): restructure.
+namespace Client {
+
+/**
+ * @brief We translate certain exceptions thrown by TCLAP to NoServingException, for example when
+ * help is invoked. This exception is then caught further up the stack and handled.
+ *
```
Comment (htuch) [inline nighthawk/include/nighthawk/common/exception.h]: Nit: skip blank line ere and below;.

File: nighthawk/source/exe/BUILD [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    "//bazel:nighthawk.bzl",
+    "nighthawk_cc_platform_dep",
+)
+load(
+    "@envoy//bazel:envoy_build_system.bzl",
+    "envoy_cc_binary",
+    "envoy_cc_library",
+    "envoy_cc_platform_dep",
+    "envoy_cc_posix_library",
+    "envoy_cc_win32_library",
```
Comment (htuch) [inline nighthawk/source/exe/BUILD]: We don't need a bunch of these I think.

File: nighthawk/source/exe/BUILD [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    "envoy_package",
+)
+load("@envoy//api/bazel:api_build_system.bzl", "api_proto_library_internal")
+load("@envoy//bazel:repositories.bzl", "PPC_SKIP_TARGETS")
```
Comment (htuch) [inline nighthawk/source/exe/BUILD]: Not needed.

**Comment by oschaaf** [comment]
@htuch all comments resolved, CI passing with clang.
I'll take CI for another spin to see if that CI is now passing more consistently.
Please take another look to see if this is good to go.
/retest

File: ci/do_ci.sh [hunk 9]
```diff
@@ -1,4 +1,40 @@
...
+
+# TODO(oschaaf): To avoid OOM kicking in, we throttle resources here. Revisit this later
+# to see how this was finally resolved in Envoy's code base. There is a TODO for when
+# when a later bazel version is deployed in CI here:
+# https://github.com/lizan/envoy/blob/2eb772ac7518c8fbf2a8c7acbc1bf89e548d9c86/ci/do_ci.sh#L86
+[ -z "$CIRCLECI" ] || export BAZEL_BUILD_OPTIONS="${BAZEL_BUILD_OPTIONS} --local_resources=4096,2,1"
+[ -z "$CIRCLECI" ] || export BAZEL_TEST_OPTIONS="${BAZEL_TEST_OPTIONS} --local_resources=4096,2,1 --local_test_jobs=4"
+[ -z "$CIRCLECI" ] || export PATH=/usr/lib/llvm-7/bin:$PATH
+[ -z "$CIRCLECI" ] || export CC=clang
+[ -z "$CIRCLECI" ] || export CXX=clang++
```
Comment (htuch) [inline ci/do_ci.sh]: Nit: this could all be under a single `if [ -z "$CIRCLECI" ]` block.

File: nighthawk/bazel/nighthawk.bzl [hunk 10]
```diff
@@ -0,0 +1,6 @@
+def nighthawk_cc_platform_dep(name):
+    return select({
+        "@envoy//bazel:windows_x86_64": [name + "_win32"],
+        "//conditions:default": [name + "_posix"],
+    })
```
Comment (htuch) [inline nighthawk/bazel/nighthawk.bzl]: Nit: do we need this?

File: nighthawk/test/frequency_test.cc [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+#include "nighthawk/source/common/frequency.h"
+
+namespace Nighthawk {
+
+class FrequencyTest : public testing::Test {
+
+public:
+  FrequencyTest() {}
+  void SetUp() {}
+  void TearDown() {}
```
Comment (htuch) [inline nighthawk/test/frequency_test.cc]: Nit: none of these should be needed if they are just the default empty bodies.

**Comment by oschaaf** [comment]
@htuch comments addressed. I have this skeleton plus running `clang-tidy` in CI [staged for a follow up](https://github.com/envoyproxy/envoy-perf/compare/master...We-Amp:nighthawk-skeleton-clang-tidy?expand=1).
Shall we follow up with trying to land that? 
(It's running clean on my machine, the test is set up to do full checking).

**Comment by htuch** [comment]
@oschaaf yep, let's look at landing that next.

**Pull request merged**